### ISSUE TITLE
Add JetBrains Editor Support

### ISSUE THREAD

**Issue by invakid404** [BoundaryML/baml#889]
Hi! I use JetBrains IDEs (WebStorm, PyCharm, RubyMine) please add support.

**Comment by hellovai** [comment]
We've had this request quite a bit, I'll take a look next week and report back with how much work we think this is.

@invakid404 would it work for you if we did something like this in the meanwhile:

The ability to `baml-cli playground --port 3000` to spawn a webui synced to your local filesystem matching [promptfiddle.com](promptfiddle.com)?

**Comment by invakid404** [comment]
> We've had this request quite a bit, I'll take a look next week and report back with how much work we think this is.
> 
> @invakid404 would it work for you if we did something like this in the meanwhile:
> 
> We offer a `baml-cli playground --port 3000` command which would spawn a webui synced to your local filesystem matching [promptfiddle.com](promptfiddle.com)?

I would personally be fine with a playground outside the IDE, yes. I still think a basic BAML plugin for JetBrains that adds language support (syntax highlighting, LSP) and no in-IDE playground would still be very useful so that I can at least read BAML files locally and click around.

**Comment by sxlijin** [comment]
Awesome to hear! I've filed #960 as a separate task, and we'll make sure to update you when we finish that.

In the meantime, we'll keep this issue open for other folks who want Jetbrains support to let us know. Some research ntoes about what we'd have to do:

- Jetbrains custom language support docs are [here](https://plugins.jetbrains.com/docs/intellij/custom-language-support.html)
- quick skim: Jetbrains doesn't support textmate/lezer grammars, it looks like we'd have to expose more out of the LSP
- showing the preview would require some rearchitecture, since it's unclear how we'd embed the webview. could use [JCEF](https://plugins.jetbrains.com/docs/intellij/jcef.html#enabling-jcef) (embedded chromium), but that's an experimental feature.

**Comment by invakid404** [comment]
> Awesome to hear! I've filed #960 as a separate task, and we'll make sure to update you when we finish that.
> 
> In the meantime, we'll keep this issue open for other folks who want Jetbrains support to let us know. Some research ntoes about what we'd have to do:
> 
> * Jetbrains custom language support docs are [here](https://plugins.jetbrains.com/docs/intellij/custom-language-support.html)
> * quick skim: Jetbrains doesn't support textmate/lezer grammars, it looks like we'd have to expose more out of the LSP
> * showing the preview would require some rearchitecture, since it's unclear how we'd embed the webview. could use [JCEF](https://plugins.jetbrains.com/docs/intellij/jcef.html#enabling-jcef) (embedded chromium), but that's an experimental feature.

JetBrains IDEs support TextMate grammars via the TextMate plugin bundled with all JetBrains IDEs. Plugins can provide TextMate bundles by implementing [TextMateBundleProvider](https://github.com/JetBrains/intellij-community/blob/idea/242.21829.142/plugins/textmate/src/org/jetbrains/plugins/textmate/api/TextMateBundleProvider.kt).

**Comment by angelozerr** [comment]
For LSP tropic, jetbrains provides an lsp support for ultimate which is not free.

If you want to have a free lsp support, please try https://github.com/redhat-developer/lsp4ij which is a free LSP support IntelliJ plugin.

If you are interested to evaluate it, see the following documentation:

 * the [LSP support from LSP4IJ](https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md) 
 * the [Developer Guide](https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md)
 * the [LSP console](https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserGuide.md)

And if you want to evaluate quickly LSP4IJ without developping a plugin but just filling in a settings the start command of your ls, please see https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserDefinedLanguageServer.md

I suggest that you read article https://idetools.dev/blog/lsp4ij-announcement/ to have an overview of LSP4IJ

**Comment by glofru** [comment]
Bump this issue up. I understand why you preferred to implement VSCode first, but IntelliJ products are [50+% of the Python development market](https://www.jetbrains.com/lp/devecosystem-2023/python/#python_ide). You may consider it for a wider developer adoption.

**Comment by kemiller** [comment]
Adding one more +1 to this.  Ruby support in RubyMine is *vastly* better than anything you can get in VSCode and its derivatives. Currently I go back and forth to Windsurf but it's not ideal.  Just syntax files + a cli playground that watches baml files and auto-compiles them would be good enough.

**Comment by Elijas** [comment]
+1 for our team too
(Pycharm professional)

**Comment by angelozerr** [comment]
If someone can explain me how to install and start the baml language server, I can share with you a simple template for LSP4IJ and you could consume baml language server inside IntelliJ.

**Comment by hellovai** [comment]
funny enough, @sxlijin took this on for his friday. We'll find out how far he got on monday!

![Image](https://github.com/user-attachments/assets/6fbea518-f711-4a21-8f62-902e8da1bffa)

@angelozerr fyi the command is:

```
baml-cli lsp --port <PORT>
```

**Comment by angelozerr** [comment]
> funny enough, [@sxlijin](https://github.com/sxlijin) took this on for his friday. We'll find out how far he got on monday!
> 
> ![Image](https://github.com/user-attachments/assets/6fbea518-f711-4a21-8f62-902e8da1bffa)
> 
> [@angelozerr](https://github.com/angelozerr) fyi the command is:
> 
> ```
> baml-cli lsp --port <PORT>
> ```

Thanks! And do you know how to install it?

Oh that's bad for LSP4IJ, for the moment we don't support socket, only stdio. I need to work on this support.

**Comment by angelozerr** [comment]
After testing baml, it seems that it is working  without port (socket mode) and can support stdio mode.

In few minutes you can configure IJ:

 * to use textmate and language-configuration from vscode baml to have syntax coloration
 * register the LSP baml language server with LSP4IJ (without developping an IJ plugin).

Here a demo in IntelliJ:

![Image](https://github.com/user-attachments/assets/35fdb183-4f03-4a71-bfe1-4187ce9e3c79)

It is very basic. If you want to have the same features like vscode, an IJ plugin is required to implement the custom LSP command.

For instance codelens are displayed:

![Image](https://github.com/user-attachments/assets/74513185-fd5d-4cee-a8ed-960bcdacccd3)

But when you click on it it does nothing because the LSP command is not defined in the IJ client.

I don't know if it will interest some Intellij users with those basic features.

@hellovai I suggest that you manage vscode snippet like class on language server side, in order to any LSP client can benefit them.

**Comment by angelozerr** [comment]
@hellovai I suggest also you use LSP standard response to cover the most features with any LSP client like LSP4IJ.

 * snippet support
 * when you need to generate file (when a baml file is saved), use https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_applyEdit

**Comment by hellovai** [comment]
@angelozerr this is amazing. thanks for sharing this. I'll report back with the team. Since @sxlijin has been going down the road of a proper extension, I think what you shared is a great middle-ground to support the intelij ecosystem.

Are there simple instructions on how to run this? (Otherwise, I can debug this later myself) 
Apologies for the basic questions, have never used intelij ecosystem myself :)

**Comment by angelozerr** [comment]
> [@angelozerr](https://github.com/angelozerr) this is amazing. thanks for sharing this. I'll report back with the team. Since [@sxlijin](https://github.com/sxlijin) has been going down the road of a proper extension, I think what you shared is a great middle-ground to support the intelij ecosystem.
> 
> Are there simple instructions on how to run this? (Otherwise, I can debug this later myself) Apologies for the basic questions, have never used intelij ecosystem myself :)

 * Install IntelliJ
 * Do Ctrl+Alt+S (for windows) to open IntelliJ settings, choose Plugins and search LSP4IJ

![Image](https://github.com/user-attachments/assets/67a53c33-44b6-47c2-8929-11059a0fae28)

 * install LSP4IJ
 * restart IntellIJ you should see (empty) language server view

![Image](https://github.com/user-attachments/assets/ece410bb-fa19-43d8-a234-80ef628272aa)

# LSP support

LSP4IJ provides an LSP client support for IntelliJ, the best thing is to integrate the language server with an IJ plugin, see https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md to provide custom command (like you are using in your language server)

But LSP4IJ provides also the capability to integrate any LSP language server with just settings, see https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserDefinedLanguageServer.md (I repeat me, it is not the poper solution, it i sbetter to write an IJ plugin to manage start command of the language server (according the OS), provide custom LSP command), etc.

You could define Baml language server at hand, you can import template https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserDefinedLanguageServer.md#custom-template

Here the instruction:

 * donwload this zip which defines the start command for baml language server (it is for windows, but you will need to adjust for other OS):

[lsp4ij-export20250519.zip](https://github.com/user-attachments/files/20300383/lsp4ij-export20250519.zip)

 * unzip it
 * create a New Language Server https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserDefinedLanguageServer.md#new-language-server-dialog
 * import the unzipped template https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserDefinedLanguageServer.md#custom-template

![Image](https://github.com/user-attachments/assets/97b1df87-701c-4260-baed-df371d3291ca)

 * after click on OK you should see Balm language server 

![Image](https://github.com/user-attachments/assets/885b9582-cb67-4e5e-9199-a2af8b68131f)

The command for windows is `cmd /C $PROJECT_DIR$/baml-cli.exe lsp` `$PROJECT_DIR$` is a special IntelliJ variable which resolve the project root. You can see on the bottom the resolved path:

![Image](https://github.com/user-attachments/assets/2a641280-1570-4f5b-a3fb-482cefb87759)

In my case I have copied/pasted balm-cli.exe in my project root dir

![Image](https://github.com/user-attachments/assets/5ca0a30d-25f7-4fdb-be62-de4a07ec6e0a)

 * for other OS you should adjust command like

`sh -c $PROJECT_DIR$/baml-cli lsp`

I don't know if it will work, just give you some hint.

* the template contains some JSON files which explains how to associate file with baml language server in mappings section

![Image](https://github.com/user-attachments/assets/510fcdc6-8375-4688-85c0-65f5d61d3fda) You can add , remove update them

Open a baml file and if command can be started you should see the feature like diagnostics, definition like my demo.


# Syntax coloration

IntelliJ by default can support TextMate and load textmate and language-configuration.json from a package.json of vscode.

 * Clone baml repository
 * Open Settings / TextMate

![Image](https://github.com/user-attachments/assets/4910fd34-4f29-4491-b4d6-543903fbeeca)

 * click on + (Add) and select https://github.com/BoundaryML/baml/tree/canary/typescript/vscode-ext/packages which contains the package.json

<img width="377" alt="Image" src="https://github.com/user-attachments/assets/fd073901-bb0a-42cc-a81d-7e8e41f37f43" />

You shoudl have syntax coloration and language configuration support (autoclose, etc)

**Comment by sxlijin** [comment]
Sorry for the radio silence from me - Vaibhav's been answering because I'm traveling right now - but this looks really cool! I'm very excited to have a way to offer LSP support to folks using the community editions of the various Jetbrains IDEs, so I'll absolutely look into making our plugin depend on LSP4IJ.

**Comment by angelozerr** [comment]
No problem.

I am very glzd thzt yiu could be interested by lsp4ij

I am trying to do my best to use LSP4IJ as lsp support since jetbrains provide a non free lsp support.

There are more and more plugins based on Lsp4ij https://github.com/redhat-developer/lsp4ij?tab=readme-ov-file#who-is-using-lsp4ij and lsp4ij community is growing up and have more and more contributions 

Hope someone could be interested to implement an ij plugin for Balm with lsp4ij.

I am working on server installer support. It will help to download znd install baml easily and configure the command zccording the os.

**Comment by sxlijin** [comment]
FYI re installer - we'll probably want to sync with you on that, because we have some custom logic in our vscode installer w.r.t. how we choose the version that we invoke. I'm not sure how much of that we could offload to lsp4ij, as much as I'd like to.

**Comment by angelozerr** [comment]
Yes sure!

If you implement with an IJ plugin your server it aleady exists an installer api that you can implement as you wish. See https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPApi.md#language-server-installer

For now I am working on installer to declare installation steps with json.

**Comment by egol** [comment]
#2001
Basic support for Jetbrains now merged into the main branch. Including the playground!
Note: waiting on release with the updated LSP in order for the automatic downloading to work
